sessions/return-oriented-programming/README.md

```diff
@@ -0,0 +1,810 @@
+- At `pop ebx`, `0x61626364` is loaded into `ebx` and the stack is increased
+again.
+- At `leave`, two things actually happen: `mov esp, ebp; pop ebp`. So the stack
+frame is decreased to the previous one (pointed by `ebp`) and `ebp` is updated
+to the one before that. So `esp` will now be the old `ebp + 4`.
+- At `ret`, the code flow will go to the instruction pointed to by `ebp+4`. This
+implies that execution will not go to `0xAABBCCDD` but to some other address
+that may or may not be in our control (depending on how much we can overflow on
+the stack). If it is in our control we can overwrite that address with the rest
+of the ROP chain.
+
+### Changing Register Values
+We have now seen how gadgets can be useful if we want the CPU to achieve a
+certain state. This is particularly useful on other architectures such as ARM
+and x86_64 where functions do not take parameters from the stack but from
+registers. As an example, if we want to call `f1(0xAB, 0xCD, 0xEF)` on x86_64 we
+first need to know the calling convention for the first three parameters (the
+convention for placing the rest of the parameters can be found in
+[table here](https://en.wikipedia.org/wiki/X86_calling_conventions#x86-64_calling_conventions)):
+```
+1st param: RDI
+2nd param: RSI
+3rd param: RDX
+```
+
+Now we need to find gadgets for each of these parameters. Let's assume these 2
+scenarios: Scenario 1:
+```
+0x400124:  pop rdi; pop rsi; ret
+0x400235:  pop rdx; ret
+0x400440:  f1()
+
+Payload:
+RET + 0x00:   0x400124
+RET + 0x08:   val of RDI (0xAB)
+RET + 0x10:   val of RSI (0xCD)
+RET + 0x18:   0x400235
+RET + 0x20:   val of RDX
+RET + 0x28:   f1
+```
+
+Scenario 2:
+```
+0x400125:  pop rdi; ret
+0x400252:  pop rsi; ret
+0x400235:  pop rdx; ret
+0x400440:  f1()
+
+Payload:
+RET + 0x00:   0x400125
+RET + 0x08:   val of RDI (0xAB)
+RET + 0x10:   0x400252
+RET + 0x18:   val of RSI (0xCD)
+RET + 0x20:   0x400235 
+RET + 0x28:   val of RDX
+RET + 0x30:   f1
+```
+Notice that because the architecture is 64 bits wide, the values on the stack
+are not dwords but qwords (quad words: 8 bytes wide). Thus, the offsets between
+the values in the payload are 8, instead of 4 (as they would be on a 32-bit
+architecture).
+
+### Clearing the Stack
+The second use of gadgets is to clear the stack. Remember the issue we had in
+the [Motivation](#motivation) section? Let's solve it using gadgets. We need to call
+`f1(0xAB, 0xCD)` and then `f2(0xEF, 0x42)`. Our initial solution was:
+```
+RET + 0x00:   addr of f1
+RET + 0x04:   addr of f2 (return address after f1 finishes)
+RET + 0x08:   0xAB (param1 of f1)  
+RET + 0x0c:   0xCD (param2 of f1)  but this should also be 0xEF (param1 of f2)
+RET + 0x10:   0x42 (param2 of f2) 
+```
+
+The problem is that those parameters of `f1` are getting in the way of calling
+`f2`. We need to find a `pop pop ret` gadget. The actual registers are not
+important, as we only need to clear 2 values from the stack.

```

Comment:
Signal that you are going back to `x32` to eliminate any potential confusion about architecture.